db/sqlinterfaces.c

```diff
@@ -4621,60 +4621,52 @@ __func__, cnonce);
 
 */
 
-int check_active_appsock_connections(struct sqlclntstate *clnt)
-{
-    int max_appsock_conns =
-        bdb_attr_get(thedb->bdb_attr, BDB_ATTR_MAXAPPSOCKSLIMIT);
-    if (active_appsock_conns > max_appsock_conns) {
-        int num_retry = 0;
-        int rc = -1;
-    retry:
-        Pthread_mutex_lock(&clnt_lk);
-        Pthread_mutex_lock(&appsock_conn_lk);
-        num_retry++;
-        if (active_appsock_conns <= max_appsock_conns) {
-            Pthread_mutex_unlock(&appsock_conn_lk);
-            Pthread_mutex_unlock(&clnt_lk);
-            return 0;
-        }
-        struct sqlclntstate *lru_clnt = listc_rtl(&clntlist);
-        listc_abl(&clntlist, lru_clnt);
-        while (lru_clnt != clnt) {
-            if (lru_clnt->done && !in_client_trans(lru_clnt)) {
-                lru_clnt->statement_timedout = 1; /* disallow any new query */
-                break;
-            }
-            lru_clnt = listc_rtl(&clntlist);
-            listc_abl(&clntlist, lru_clnt);
-        }
 
-        if (lru_clnt == clnt || !lru_clnt->done) {
-            /* All clients have transactions, wait for 1 second */
-            if (num_retry <= 5) {
-                Pthread_mutex_unlock(&appsock_conn_lk);
-                Pthread_mutex_unlock(&clnt_lk);
-                sleep(1);
-                goto retry;
-            }
-            rc = -1;
+static int check_done_func(void *obj) {
+    struct sqlclntstate *clnt = (struct sqlclntstate *) obj;
+    /* Don't find the client which we already killed */
+    if (clnt->done && !in_client_trans(clnt) && !clnt->statement_timedout) {
+        return 0;
+    }
+    return -1;
+}
+
+int close_lru_client(int curr_conns, struct sqlclntstate *clnt, int max_iteration) {
+    int rc = -1;
+    Pthread_mutex_lock(&clnt_lk);
+    struct sqlclntstate *lru_clnt = listc_findl(&clntlist, check_done_func, clnt, max_iteration);
+    if (lru_clnt) {
+        if (lru_clnt == clnt) {
+             /* This is a new client however its one of the  first 'max_iteration' connections. */
+             rc = 0;
         } else {
+            lru_clnt->statement_timedout = 1; /* disallow any new query */
             int fd = sbuf2fileno(lru_clnt->sb);
-            if (lru_clnt->done) {
-                // lru_clnt->statement_timedout = 1; already done
-                shutdown(fd, SHUT_RD);
-                logmsg(
-                    LOGMSG_WARN,
-                    "%s: Closing least recently used connection fd %d , total "
-                    "%d \n",
-                    __func__, fd, active_appsock_conns - 1);
-                rc = 0;
-            }
+            shutdown(fd, SHUT_RD);
```

Comment:
We always call this with clnt == NULL, so I don't think we can rely on this not finding our own connection.

Reply:
Used with clnt here.. to make it work with other plugins https://github.com/bloomberg/comdb2/pull/2396/files#diff-4932981bbf2c2bd6bf0267d1a6d39173R4664